### DISCUSSION TITLE
'NoneType' object has no attribute 'coordinates'

### DISCUSSION THREAD

**Discussion by JonasLi-19** [RMeli/spyrmsd#96] [Q&A] [answered]
I have read the tutorial but there are something wrong with my code:

```python
from spyrmsd import io, rmsd, molecule

...
            ref_pose = os.path.join(crystal_dir, pdb_id, f'{pdb_id}_ligand.sdf')
            ref = io.loadmol(ref_pose)
            mol = io.loadmol(new_filepath)

            ref = ref.strip()
            mol = mol.strip()

            coords_ref = ref.coordinates
            anum_ref = ref.atomicnums
            adj_ref = ref.adjacency_matrix

            coords = mol.coordinates
            anum = mol.atomicnums
            adj = mol.adjacency_matrix

            symm_rmsd_value = rmsd.symmrmsd(
                coords_ref,
                coords,
                anum_ref,
                anum,
                adj_ref,
                adj,
            )
```
Then it raises ERROR:

```
Traceback (most recent call last):
  File "analysis_docked_poses_score_to_csv.py", line 70, in <module>
    coords_ref = ref.coordinates
AttributeError: 'NoneType' object has no attribute 'coordinates'
```

**Comment by RMeli** [comment]
This error indicates that the `ref` object is `None`. This happens if the molecule is not properly loaded. What are you using as a backend? `rdkit` does return `None` if a molecule can't be parsed/sanitized. Do you get any warnings/error from RDKit/Open Babel when loading the molecule?

**Reply by JonasLi-19** [reply]
I thought if I use io.load, then I don't need openbabel/RDKit?
```
The spyrmsd.io module provides functions to easily load molecules from a file and to transform them into a spyrmsd.molecule.Molecule object:

ref = io.loadmol("molecules/1a4k_ligand.sdf")
```

This is the code I tried using `io.load()`, without openbabel/rdkit:
```python
from spyrmsd import io, rmsd, molecule

...
            ref_pose = os.path.join(crystal_dir, pdb_id, f'{pdb_id}_ligand.sdf')
            ref = io.loadmol(ref_pose)
            mol = io.loadmol(new_filepath)

            ref = ref.strip()
            mol = mol.strip()

            coords_ref = ref.coordinates
            anum_ref = ref.atomicnums
            adj_ref = ref.adjacency_matrix

            coords = mol.coordinates
            anum = mol.atomicnums
            adj = mol.adjacency_matrix
```

**Reply by JonasLi-19** [reply]
And this link is the code I tried with RDKit, but still error:  https://github.com/RMeli/spyrmsd/discussions/97#discussioncomment-7984976
There is indeed some warning with the molecule:
```
[20:23:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.
[20:23:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.
Traceback (most recent call last):
  File "analysis_docked_poses_score_to_csv.py", line 82, in <module>
    coords_ref = ref_molecule.coordinates
AttributeError: 'NoneType' object has no attribute 'coordinates'
```

**Reply by RMeli** [reply]
`spyrmsd.io` still needs either RDKit or OpenBabel to be installed. My personal preference is for RDKit, which has stricter checks. `spyrmsd` does not implement parsing molecular files, that's what RDKit and Open Babel do well already.

Once you have RDKit installed, you can use `io.load` to load the molecules, without having to use RDKit directly yourself (as you did in https://github.com/RMeli/spyrmsd/discussions/97#discussioncomment-7984976). So, once RDKit is installed the code should work. The error you see is likely caused by issues with your file, which can't be properly parsed by RDKit.

I haven't seen this issue before, but I suspect that coordinates are not properly included given that the molecule is tagged as 2D?

**Reply by RMeli** [reply, marked as answer]
BTW, if you have `spyrmsd` molecules already, you can use the `rmsdwrapper` function instead of passing the information directly to `symmrmsd`. I should document this better for sure.

```python
from spyrmsd import io, rmsd, molecule

...
            ref_pose = os.path.join(crystal_dir, pdb_id, f'{pdb_id}_ligand.sdf')
            ref = io.loadmol(ref_pose)
            mol = io.loadmol(new_filepath)

            reds = rmsdwrapper(ref, mol) # By default, strips hydrogens and use symmetry-correction
```

https://github.com/RMeli/spyrmsd/blob/8d3c540c2f6c822b89aa32f21e0533e84059a665/spyrmsd/rmsd.py#L300-L308

**Comment by RMeli** [comment]
@Dadiao-shuai, I got an e-mail notification with your question, but I can't see it anymore on GitHub. Did you remove it? (Please don't. It's a bit confusing for me, and every question is valuable for the community.)

---

The RDKit error
```
ERROR: Could not sanitise molecule ending on line 102
ERROR: Explicit valence for atom # 6 C, 6, is greater than permitted
```
means that RDKit can't sanitise the molecule because of problems with your input file (specifically, there is a carbon with 6 bonds). This leads to the loaded molecule being `None`, hence the `NoneType` error.

I'll try to provide a better error message when molecules can't be loaded by RDKit or OpenBabel instead of having `None`s around.
